### PR TITLE
[5.5] Standerize response object

### PR DESCRIPTION
**Pull request by themsaid**
The aim of this PR is to make sure the return value of `Router::prepareResponse` to always be an instance of `Illuminate\Http\Response`, this way developers can interact with the response object in their middleware knowing what type it is.

It would solve an issue like in: https://github.com/laravel/passport/pull/474

### CODE REVIEW

File: tests/Routing/RoutingRouteTest.php [hunk 1]
```diff
@@ -1403,8 +1403,12 @@ public function testJsonResponseIsReturned()
...
 
         $response = $router->dispatch(Request::create('foo/bar', 'GET'));
-        $this->assertNotInstanceOf(\Illuminate\Http\Response::class, $response);
-        $this->assertInstanceOf(\Illuminate\Http\JsonResponse::class, $response);
+
+        $response->setEncodingOptions(0);
+
+        $this->assertInstanceOf(\Illuminate\Http\Response::class, $response);
+        $this->assertInstanceOf(\Illuminate\Http\JsonResponse::class, $response->baseResponse);
+        $this->assertEquals(0, $response->getEncodingOptions());
```
Comment (lucasmichot) [inline tests/Routing/RoutingRouteTest.php]: `$this->assertEquals($response->getData(true), $response->baseResponse->getData(true));`fails unfortunatelly

File: src/Illuminate/Routing/Router.php [hunk 2]
```diff
@@ -674,6 +677,21 @@ public static function prepareResponse($request, $response)
...
     /**
+     * Create an illuminate response from a given base response object.
+     *
+     * @param  mixed  $response
+     * @return \Illuminate\Http\Response
+     */
+    private static function createResponseFromBase($response)
+    {
+        return (new Response(
+            $response->getContent(),
```
Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: So we lose will lose the benefits of `getOriginalContent` at this point.

I believe we should try first to take `getOriginalContent` if the method exists, then fallback to `getContent`

File: src/Illuminate/Routing/Router.php [hunk 3]
```diff
@@ -674,6 +677,21 @@ public static function prepareResponse($request, $response)
...
+            $response->getStatusCode(),
+            $response->headers->all()
+        ))->setBaseResponse($response);
```
Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: Shouldn't we set the original Response `$exception` here ?

File: tests/Routing/RoutingRouteTest.php (hunk 1 above)
Comment (themsaid) [inline tests/Routing/RoutingRouteTest.php]: took care of that.

File: src/Illuminate/Routing/Router.php (hunk 2 above)
Comment (themsaid) [inline src/Illuminate/Routing/Router.php]: What benefits do we lose? trying to understand a use case.

File: tests/Routing/RoutingRouteTest.php (hunk 1 above)
Comment (lucasmichot) [inline tests/Routing/RoutingRouteTest.php]: 👍

File: src/Illuminate/Routing/Router.php (hunk 3 above)
Comment (themsaid) [inline src/Illuminate/Routing/Router.php]: Took care of that too.

File: src/Illuminate/Routing/Router.php (hunk 2 above)
Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: @themsaid, with `return response()->json($model);` in my controller, I expect to be able to use `$model = $response->getOriginalContent();` in a middleware (to automatically set a `Last-Modified` header for instance).

Comment (themsaid) [inline src/Illuminate/Routing/Router.php]: `$response->baseResponse ? $response->baseResponse->getOriginalContent() : ...` ?

Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: @themsaid , why not just
```php
$content = method_exists($response, 'getOriginalContent')
    ? $response->getOriginalContent()
    : $response->getContent();
```

Comment (themsaid) [inline src/Illuminate/Routing/Router.php]: I think the result of conversion to JSON in `JsonResponse` is different than in `Response` so I imagine using the final output of the original response as the output of the new response is better, at the end of the day I think it's an edge case that you can easily get around by using the baseResponse.

Maybe I'm confused and there are more use cases, but I don't think reconverting the original content inside `Response` makes sense.

Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: Hmm, actually we might really miss something there.
Response will just convert the originalContent just like it was done in JsonResponse.

Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: Also why would you access it via $response->getOriginalContent and  $response->baseResponse->getOriginalContent in another case

Comment (lucasmichot) [inline src/Illuminate/Routing/Router.php]: I see your point, but it seems inconsistant

**Comment by taylorotwell** [comment]
I'm not particularly a fan of having this base response property.

**Comment by rscafi** [comment]
So, were it leave us? And laravel/passport#474

**Pull request closed without merging**